---
title: "Genes associated with prostate cancer gleason score"
author: "Ali Ekrem Yesilkanal"
date: "November 20, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal

Prostate cancer is one of the most aggressive and difficult to treat cancer types. Using TCGA gene expression data for prostate cancer along with clinical metadata, please perform the following analysis:

Starting from PRAD.uncv2.mRNAseq_raw_counts.txt from within gene expression file listed above, using limma-voom analysis identify the genes most associated with Radical Prostatectomy Gleason Score for Prostate Cancer metadata variable. In addition to providing working code for the analysis please answer the following questions:

1. What are the most important considerations when analyzing raw counts for RNA-Seq data?

2. What covariate metadata variables - if any - did you include in your analysis and why? Note that not all the metadata variables are shown by default on cbioportal and others can be toggled on.

3. If you had additional time what additional forms of analysis would you perform to isolate the genes most associated with Gleason score?

## Analysis

Most important considerations when analyzing raw counts is the use of appropriate normalization method that will allow for accurate comparisons across samples. Raw read counts are affected by gene length, library size / sequencing depth, and underlying expression differences between tissues. Longer genes will have higher chances of matching reads during the mapping process, therefore biasing read counts towards long genes. Larger libraries or higher sequence depth will increase read counts simply because there are more transcripts being matched to the genome. However, if there are differences between the library sizes across samples, this will affect downstream statistical analyses such as differential gene expression. Similarly, when the sample comes from a different origin that has a drastically different expression profile (for example from diffirent tissues or patients), highly and differentially expressed genes across samples can influence statistical comparisons. 

One common approach to adjust for these differences is to scale reads for the library size. Methods such as reads per kilobase per million mapped (RPKM), fragments per kilobase of exon model per million mapped reads (FPKM), and transcripts per million (TPM) are commonly used methods for within-sample normalization. These methods are good for comparing expression level differences between genes within the same sample. They also perform decently in studies where the samples being compared are not from heterogeneous backgrounds (for example, the samples are coming from the same tissue of origin, and replicates are used). However, if there are drastic differences in the underlying transcriptomic profiles of the samples (such as the case when the samples are coming from diffirent tisses or organs, or from different individuals with different genetic backgrounds), methods such as TMM or DEseq2 are more appropriate. These methods adjust for heterogeneous transcript distributions by ignoring highly expressed or highly variable genes. 

For the following analysis of the TCGA prostate cancer dataser, we will be using the TMM method using `edgeR`'s `calcNormFactors()` function, followed by `limma-voom` for the differential expression analysis.



```{r}
# Load necessary packages
library(limma)
library(edgeR)
library(Amelia)
library(MASS)

# Load the raw expression count data and the clinical metadata
dat.raw <- read.delim("PRAD.uncv2.mRNAseq_raw_counts.txt", na.strings=c("", " ", NA))
dat.clin <- read.delim("prad_tcga_clinical_data.tsv", na.strings=c("", " ", NA))

```

Take a look at the Gleason score data structure.

```{r}
class(dat.clin$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer)
table(dat.clin$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer) # 5 Gleason scores ranging between 6-10

```


In our analysis, we want to use expression data from samples that have associated clinical meta data. Therefore, the first step is to subset the expression data frame for the sample names listed in the clinical metadata.


```{r}
# Replace the "." with "-" within the sample ID names so they match the clinical metadata sample IDs
temp_ID <- gsub("\\.", "-", colnames(dat.raw)[-1])
colnames(dat.raw)[-1] <- temp_ID

# Consider expression data from primary cancer biopsies only (e.g. Sample IDs that end with "-01")
dat.raw2 <- dat.raw[,colnames(dat.raw) %in% dat.clin$Sample.ID]
rownames(dat.raw2) <- dat.raw$HYBRIDIZATION.R # add gene names as the row names for the expression matrix
dat.raw2[1:6,1:6]
```

Check the dimensions of the count data to make sure it matches the sample numbers in the clinical data.

```{r}
dim(dat.raw2)
```
There are 498 columns (sample IDs) in the expression data, but 499 sample IDs in the clinical data. 1 sample does not seem to have expression data even though it has clinical metadata. Find the missing sample and remove it from the clinical metadata matrix.

```{r}
dat.clin$Sample.ID[dat.clin$Sample.ID %in% colnames(dat.raw) == FALSE]
"TCGA-KC-A4BO-01" %in% colnames(dat.raw)
```
The Sample "TCGA-KC-A4BO-01" does not seem to have expression data

```{r}
# Remove TCGA-KC-A4BO-01" from dat.clin1
dat.clin1 <- dat.clin[-which(dat.clin$Sample.ID == "TCGA-KC-A4BO-01"),]
```

One last thing to check before starting voom analysis is that each sample in the clinical data have a gleason score.
```{r}
table(is.na(dat.clin1$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer))
```

There are no missing values in the Gleason score column.

Now that we have all the samples with both clinical metadata and the expression data associated with them, we can start processing the raw counts using `limma-voom`. 

The Gleason score data in the clinical dataframe are of the class "integer'. But since there are only 5 different Gleason scores in the entire dataset, I decided to treat each Gleason score as a different categorical variable.

```{r}
# Create a desing object for normalization
G <- as.factor(dat.clin1$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer)
design <- model.matrix(~0 + G)
head(design)
```

Make a DGEList object form the raw count data, filter out the genes that have low counts, and normalize using TMM method.

```{r}
# Create a DGEList object using the expression data
dge <- DGEList(counts=dat.raw2)

# Filter out genes with low counts
keep <- filterByExpr(dge, design)
dge <- dge[keep,, keep.lib.sizes = FALSE]

# Scale normalization and the TMM normalization
dge <- calcNormFactors(dge)

# Perform voom 
v <- voom(dge, design, plot=TRUE)

```

Since there are 5 categories, there are many pairings for the DE analysis. I am most interested in the genes that change with increasing Gleason Score. Clinically, Gleason score of 6 is conisdered a low grade, slow-growing, low-risk tumor. Therefore, I decided to make my DE comparisons relative to the Gleason 6 group. For example, "G10-G6" Signifies the contrast between Gleason 10 and Gleason 6.

```{r}
# Differential expression between different gleason score groups
fit <- lmFit(v, design)
contrast.matrix <- makeContrasts(G10-G6, G9-G6, G8-G6, G7-G6, levels=design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

results <- decideTests(fit2)

vennDiagram(results)
```


```{r}
# Top 20 genes associated with gleason score
topTableF(fit2, adjust.method = "BH", number=20) 
```

For example, FAM72B increases and SLC7A4 decreases significantly as the gleason score increases. We can see this by making a boxplot of FAM72B and SLC7A4 expression across Gleason score categories.

```{r}
# Visualize expression of FA2H and CD163 genes ~ gleason score
temp <- dat.clin1[,c("Sample.ID", "Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer")]
dge.counts.t <- t(dge$counts)
fam72b_exp <- dge.counts.t[,"FAM72B|653820"]
slc7a4_exp <- dge.counts.t[,"SLC7A4|6545"]
dat.temp <- cbind(temp, fam72b_exp, slc7a4_exp)

par(mfrow=c(1,2))
boxplot(fam72b_exp ~ Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer, 
        data = dat.temp, 
        main = "FAM72B gene ~ Gleason Score",
        xlab = "Gleason Score",
        ylab = "Normalized Expression")
boxplot(slc7a4_exp ~ Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer, 
        data = dat.temp, 
        main = "SLC7A4 gene ~ Gleason Score",
        xlab = "Gleason Score",
        ylab = "Normalized Expression")

```

Up to this point, the analysis was univariate. It is possible that the associations between Gleason Score and the gene expression levels are confounded by other covariates. In most studies, "Age" and "Sex" of the patiensts are taken as covariates and adjusted for. "Sex" will not be very helpful in this case since all prosate cancer patients are male. 

In order to look for other potential confounders, I decided to investigate the associations between Gleason Score and the other clinical metadata variables. Since the outcome data (Gleason score) has multiple categories, and the categories are somewhat ordered in nature (higher the score, more severe the outcome), I chose to use ordinal logistic regression method to look at associations between the outcome and other clinical variables. 

For the predictive variables, I will choose variables that do not have a lot of missing values. The following heatmap shows the missing values in the entire clinical metadata file. (I used the `missmap()` function from the `Amelia` package)

```{r}
missmap(dat.clin1, 
        main = "Missing values vs observed", 
        x.cex = 0.4, 
        y.cex = 0.4)
```

The heatmap above shows that there are a lot of clinical variables that are completely empty. I first selected variables that have only 10 or less missing values.

```{r}
# Summarize missing values in the clinical metadata
summary.missing <- sapply(dat.clin1,function(x) sum(is.na(x)))

# Remove clinical variables where there are more than 10 missing values
dat.clin2 <- dat.clin1[,!summary.missing > 10] 
missmap(dat.clin2, 
        main = "Missing values vs observed", 
        x.cex = 0.4, 
        y.cex = 0.4,
        y.labels = dat.clin2$Sample.ID)
```

The samples in the rows 381 and 443 are missing many clinical variables, so I decided to exlude those from the analysis for the sake of simplicity. 

```{r}
dat.clin3 <- dat.clin2[-c(381, 443),]
missmap(dat.clin3, 
        main = "Missing values vs observed", 
        x.cex = 0.4, 
        y.cex = 0.4,
        y.labels = dat.clin3$Sample.ID)
```

Some clinical variables have only 1 value across all samples, such as `Sample.Type` (1 facror, "Primary"). These variables will not have the outcome prediction, so we can remove them. 

```{r}
# Remove the variables with only one value
summary.unique <- sapply(dat.clin3, function(x) length(unique(x))) # 11 columns have only 1 variable
table(summary.unique)
dat.clin3 <- dat.clin3[,summary.unique > 1]
missmap(dat.clin3, 
        main = "Missing values vs observed", 
        x.cex = 0.4, 
        y.cex = 0.4,
        y.labels = dat.clin3$Sample.ID)
```

Looking at the remaining variables, we can see that some of them are ID columns. These will have unique factors for each sample, but they are not useful for predicting Gleason score.

```{r}
colnames(dat.clin3)

# Do not inlcude ID related columns in the model - so subset the metadata for molecular/genetic/clinical variables
dat.clin4 <- dat.clin3[,c(2:7, 9, 12:17, 20:21, 24:32)]
missmap(dat.clin4, 
        main = "Missing values vs observed", 
        x.cex = 0.4, 
        y.cex = 0.4,
        y.labels = dat.clin4$Sample.ID)
```

The remining variables still have missing values. I do not want to remove all of these variables because some of them might be associated with gleason score such as `Fraction.Genome.Altered`. Therefore, I will fill in the missing values with the mean of the values for a `numeric` variable, and with the mode of the values for a `factor` variable. 

```{r eval=FALSE}
# Investigate the data type of the variables that have missing values
class(dat.clin4$Primary.Tumor.Laterality) # factor
class(dat.clin4$Days.to.Last.Followup) # integer
class(dat.clin4$Primary.Lymph.Node.Presentation.Assessment.Ind.3) # factor
class(dat.clin4$Fraction.Genome.Altered) # numeric
class(dat.clin4$American.Joint.Committee.on.Cancer.Tumor.Stage.Code) # factor
class(dat.clin4$Initial.pathologic.diagnosis.method) # factor
class(dat.clin4$Disease.Free..Months.) # numeric
class(dat.clin4$Disease.Free.Status) # factor
class(dat.clin4$Tissue.Prospective.Collection.Indicator) # factor
class(dat.clin4$Tissue.Retrospective.Collection.Indicator) # factor
```


```{r echo=FALSE}
dat.clin4$Days.to.Last.Followup <- as.numeric(dat.clin4$Days.to.Last.Followup)
## Create a function to get mode
getmode <- function(x) {
  uniqx <- unique(x)
  uniqx[which.max(tabulate(match(x, uniqx)))]
}

dat.clin4$Primary.Tumor.Laterality[is.na(dat.clin4$Primary.Tumor.Laterality)] <- getmode(dat.clin4$Primary.Tumor.Laterality)
dat.clin4$Primary.Lymph.Node.Presentation.Assessment.Ind.3[is.na(dat.clin4$Primary.Lymph.Node.Presentation.Assessment.Ind.3)] <- getmode(dat.clin4$Primary.Lymph.Node.Presentation.Assessment.Ind.3)
dat.clin4$American.Joint.Committee.on.Cancer.Tumor.Stage.Code[is.na(dat.clin4$American.Joint.Committee.on.Cancer.Tumor.Stage.Code)] <- getmode(dat.clin4$American.Joint.Committee.on.Cancer.Tumor.Stage.Code)
dat.clin4$Initial.pathologic.diagnosis.method[is.na(dat.clin4$Initial.pathologic.diagnosis.method)] <- getmode(dat.clin4$Initial.pathologic.diagnosis.method)
dat.clin4$Disease.Free.Status[is.na(dat.clin4$Disease.Free.Status)] <- getmode(dat.clin4$Disease.Free.Status)
dat.clin4$Tissue.Prospective.Collection.Indicator[is.na(dat.clin4$Tissue.Prospective.Collection.Indicator)] <- getmode(dat.clin4$Tissue.Prospective.Collection.Indicator)
dat.clin4$Tissue.Retrospective.Collection.Indicator[is.na(dat.clin4$Tissue.Retrospective.Collection.Indicator)] <- getmode(dat.clin4$Tissue.Retrospective.Collection.Indicator)

dat.clin4$Fraction.Genome.Altered[is.na(dat.clin4$Fraction.Genome.Altered)] <- mean(dat.clin4$Fraction.Genome.Altered,na.rm=T)
dat.clin4$Days.to.Last.Followup[is.na(dat.clin4$Days.to.Last.Followup)] <- mean(dat.clin4$Days.to.Last.Followup,na.rm=T)
dat.clin4$Disease.Free..Months.[is.na(dat.clin4$Disease.Free..Months.)] <- mean(dat.clin4$Disease.Free..Months.,na.rm=T)

missmap(dat.clin4, 
         main = "Missing values vs observed", 
         x.cex = 0.4, 
         y.cex = 0.4,
         y.labels = dat.clin4$Sample.ID)
```

Now we can run the ordinal logistic regression with the Gleason score outcome:

```{r}
dat.clin4$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer <- as.factor(dat.clin4$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer)

m <- polr(Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer ~ ., 
          data=dat.clin4[,-1],
          Hess = TRUE,
          method = "logistic")

## store table
ctable <- coef(summary(m))

## calculate and store p values
p <- pnorm(abs(ctable[, "t value"]), lower.tail = FALSE) * 2

## combined table
ctable <- cbind(ctable, "p value" = p)
ctable <- as.data.frame(ctable)

## calculate odds ratios
OR <- exp(coef(m))
ctable <- cbind(ctable[-(60:63),], OR)

ctable[ctable$`p value` <= 0.05, c(4,5)]
```


Fraction of the altered genome can be a confounder because alterations in the genome affect expression profiles of the samples. Just like age, tumor stage can be a good predictor for gleason score, as late stage tumors tend to be higher grade (higher Gleason score). Finally seems like Tissue source site might have an effect on the gleason scoring. This could be either becuase of the personnelle doing the gleason scoring of the tissues, or becuase of differences in the sequencing technique (e.g. machine used, person performing the sequencing  etc.) Therefore, I decided to control for age, tumor stage, fraction of the altered genome, and tissue source site.

```{r}
# Run voom with the covariates selected above
G <- as.factor(dat.clin4$Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer)
A <- cut(dat.clin4$Diagnosis.Age, c(40,60,80), labels = c("Y", "O")) # Age below 60 as Y (young) and above 60 as O (old)
S <- as.factor(dat.clin4$American.Joint.Committee.on.Cancer.Tumor.Stage.Code)
FGA <- dat.clin4$Fraction.Genome.Altered
TSS <- as.factor(dat.clin4$Tissue.Source.Site)

design2 <- model.matrix(~ 0 + G + A + S + FGA + TSS, data=dat.clin4)
colnames(design2)

#resubset the raw data for the sample ids that we have in dat.clin4
dat.raw3 <- dat.raw2[,colnames(dat.raw2) %in% dat.clin4$Sample.ID]

# Create a DGEList object using the expression data
dge2 <- DGEList(counts=dat.raw3)

# Filter out genes with low counts
keep2 <- filterByExpr(dge2, design2)
dge2 <- dge2[keep2,, keep.lib.sizes = FALSE]

# Scale normalization and the TMM normalization
dge2 <- calcNormFactors(dge2)

v2 <- voom(dge2, design2, plot=FALSE)

fit_new <- lmFit(v2, design2)

contrast.matrix2 <- makeContrasts(G10-G6, G9-G6, G8-G6, G7-G6, levels=design2)
fit_new2 <- contrasts.fit(fit_new, contrast.matrix2)
fit_new2 <- eBayes(fit_new2)

results2 <- decideTests(fit_new2)

vennDiagram(results2)

topTableF(fit_new2, number=30) 
```

After adjusting for these clinical variables, the gene list that is significantly associated with the gleason score has changed. Even though CHRM1 gene is still on the list, it lost rank. Some of the higher ranked genes are visualized below:

```{r}
# Visualize expression of FA2H and CD163 genes ~ gleason score
temp <- dat.clin4[,c("Sample.ID", "Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer")]
dge2.counts.t <- t(dge2$counts)
fa2h_exp <- dge2.counts.t[,"FA2H|79152"]
cd163_exp <- dge2.counts.t[,"CD163|9332"]
dat.temp <- cbind(temp, fa2h_exp, cd163_exp)

par(mfrow=c(1,2))
boxplot(fa2h_exp ~ Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer, 
        data = dat.temp, 
        main = "FA2H gene ~ Gleason Score",
        xlab = "Gleason Score",
        ylab = "Normalized Expression")
boxplot(cd163_exp ~ Radical.Prostatectomy.Gleason.Score.for.Prostate.Cancer, 
        data = dat.temp, 
        main = "CD163 gene ~ Gleason Score",
        xlab = "Gleason Score",
        ylab = "Normalized Expression")
```

If I had more time, there are a few more approaches that I would try to find the genes that are most associated with the Gleason score

1- I appraoched the Gleason score variable as a multi-class problem where each score was an independent class. However, the data distribution was not equal among classes. Gleason score 10, for example, had only 3-5 samples in the dataset. Looking at the genes listed above, it is obvious that some of those genes show a drastic difference only in the G10-G6 comparison. This might be some artifact of the class size imbalance. To get around this, I could approach the problem as a two-class problem, where I designate Gleason scores 6-7 as "low grade" and Gleason scores 8-10 as "high grade". This grouping also has clinical implication becuase low Gleason score tumors are considered slow-growing (non-aggressive), and high Gleason Score tumors are fast-growing (aggressive). Finding genes associated with this binomial grouping might be biologically more meaningful. Also, statistically-speaking, two-class problems are much easier to deal with.

2- Another approach would be to use cope number variation (CNV) data or genomic data (instead of transcriptomic). We could investigate the associations between the genomic alterations such as deletions and multiplications in prostate cancer with the gleason scoring of the tumor. In the case where a significant association is observed, we could go deeper into the genes that are present in the altered region of the genome. Presumably, these genes would be drivers of the tumor aggressiveness and good predictors of the gleason score. 

```{r}
sessionInfo()
```








